programs/stake/src/stake_state.rs

```diff
@@ -1158,6 +1144,135 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
     }
 }
 
+#[derive(Clone, Debug, PartialEq)]
+enum MergeKind {
+    Inactive(Meta, u64),
+    ActivationEpoch(Meta, Stake),
+    FullyActive(Meta, Stake),
+}
+
+impl MergeKind {
+    fn meta(&self) -> &Meta {
+        match self {
+            Self::Inactive(meta, _) => meta,
+            Self::ActivationEpoch(meta, _) => meta,
+            Self::FullyActive(meta, _) => meta,
+        }
+    }
+
+    fn stake(&self) -> Option<&Stake> {
+        match self {
+            Self::Inactive(_, _) => None,
+            Self::ActivationEpoch(_, stake) => Some(stake),
+            Self::FullyActive(_, stake) => Some(stake),
+        }
+    }
+
+    fn get_if_mergeable(
+        stake_keyed_account: &KeyedAccount,
+        clock: &Clock,
+        stake_history: &StakeHistory,
+    ) -> Result<Self, InstructionError> {
+        match stake_keyed_account.state()? {
+            StakeState::Stake(meta, stake) => {
+                // stake must not be in a transient state. Transient here meaning
+                // activating or deactivating with non-zero effective stake.
+                match stake.delegation.stake_activating_and_deactivating(
```

Comment:
Would you mind adding a comment on `stake_activating_and_deactivating()` defining the return values?
Those opaque u64s make me 😢 
I wouldn't mind a comment somewhere here as well, as a reminder of what those tuples below mean.

Reply:
I want a struct there so bad

Reply:
💯 same